### DISCUSSION TITLE
Installing on Raspberry Pi

### DISCUSSION THREAD

**Discussion by physimo** [withcatai/node-llama-cpp#372] [Q&A] [answered]
Got this error when I try to install v3.0.1

```
npm ERR! ERR! OMG Process terminated: 2
npm ERR! Failed to build llama.cpp with no GPU support. Error: SpawnError: Command npm run -s cmake-js-llama -- compile --log-level warn --config Release --arch=arm64 --out localBuilds/linux-arm64 --runtime-version=20.10.0 --cmake-path "/home/physimo/llama/node_modules/node-llama-cpp/llama/xpack/xpacks/.bin/cmake" --CDGGML_METAL=OFF --CDGGML_CCACHE=OFF exited with code 1
npm ERR!     at createError (file:///home/physimo/llama/node_modules/node-llama-cpp/dist/utils/spawnCommand.js:34:20)
npm ERR!     at ChildProcess.<anonymous> (file:///home/physimo/llama/node_modules/node-llama-cpp/dist/utils/spawnCommand.js:47:24)
npm ERR!     at ChildProcess.emit (node:events:514:28)
npm ERR!     at ChildProcess._handle.onexit (node:internal/child_process:294:12)
npm ERR! SpawnError: Command npm run -s cmake-js-llama -- compile --log-level warn --config Release --arch=arm64 --out localBuilds/linux-arm64 --runtime-version=20.10.0 --cmake-path "/home/physimo/llama/node_modules/node-llama-cpp/llama/xpack/xpacks/.bin/cmake" --CDGGML_METAL=OFF --CDGGML_CCACHE=OFF exited with code 1
npm ERR!     at createError (file:///home/physimo/llama/node_modules/node-llama-cpp/dist/utils/spawnCommand.js:34:20)
npm ERR!     at ChildProcess.<anonymous> (file:///home/physimo/llama/node_modules/node-llama-cpp/dist/utils/spawnCommand.js:47:24)
npm ERR!     at ChildProcess.emit (node:events:514:28)
npm ERR!     at ChildProcess._handle.onexit (node:internal/child_process:294:12)

npm ERR! A complete log of this run can be found in: /home/physimo/.npm/_logs/2024-10-24T20_02_55_374Z-debug-0.log
```

I saw a post [like this one](https://www.tomshardware.com/how-to/create-ai-chatbot-server-on-raspberry-pi), so I thought I pretty much have to just install node-llama-cpp if I want to use NodeJS with it.

Help?

**Comment by giladgd** [comment, marked as answer]
It seems that the `npm` process exited with code `1` without printing a reason for this to the console, so there are no logs for this.

Since I see that the code is running on Linux, it may be related to the installation of nodejs, SELinux policy, or improper permissions.
There's an issue with some nodejs installations on Linux where nodejs cannot spaw other processes, and this is required for building `llama.cpp` from source.
From my experience, it usually happens when installing nodejs from snap, so if that's the case, try to uninstall it and install it from [NodeSource](https://github.com/nodesource/distributions) instead.

If you can run `npx -y node-llama-cpp inspect gpu` and attach the output of it, it would help me figure out what the issue is.

**Reply by physimo** [reply]
Ah right, I forgot to attach the full logs. I'm sorry about that.
[2024-10-25T16_28_02_808Z-debug-0.log](https://github.com/user-attachments/files/17525120/2024-10-25T16_28_02_808Z-debug-0.log)

I'm using nvm (Node Version Manager) and Node v20.10.0. Do you think I really need a clean install of NodeJS?

Also, I tried running `npx -y node-llama-cpp inspect gpu`, but it logs nothing:
```
physimo@RaspberryPi:~/Desktop/llama$ npx -y node-llama-cpp inspect gpu
npm WARN deprecated npmlog@6.0.2: This package is no longer supported.
npm WARN deprecated are-we-there-yet@3.0.1: This package is no longer supported.
npm WARN deprecated gauge@4.0.4: This package is no longer supported.
physimo@RaspberryPi:~/Desktop/llama$ 
```

**Reply by giladgd** [reply]
After inspecting the logs, it looks that your OS' `glibc` is too old to use the prebuilt binaries, and the `clang` compiler or OS itself may be too old to be able to compile the source code since it uses new cpp features that are not supported by the available compiler.

Try to run these commands and attach their output so I can figure if that's indeed the case:
```bash
cat /etc/os-release
echo "---"
cat /proc/device-tree/model
echo "---"
uname -a
echo "---"
/lib/aarch64-linux-gnu/libc.so.6
```

**Reply by physimo** [reply]
Please, take a look:

```
physimo@RaspberryPi:~/Desktop/llama$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
physimo@RaspberryPi:~/Desktop/llama$ echo "---"
---
physimo@RaspberryPi:~/Desktop/llama$ cat /proc/device-tree/model
Raspberry Pi 4 Model B Rev 1.2
physimo@RaspberryPi:~/Desktop/llama$ echo "---"
---
physimo@RaspberryPi:~/Desktop/llama$ uname -a
Linux RaspberryPi 5.4.0-1021-raspi #24-Ubuntu SMP PREEMPT Mon Oct 5 09:59:23 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
physimo@RaspberryPi:~/Desktop/llama$ echo "---"
---
physimo@RaspberryPi:~/Desktop/llama$ /lib/aarch64-linux-gnu/libc.so.6
GNU C Library (Ubuntu GLIBC 2.31-0ubuntu9.1) stable release version 2.31.
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 9.3.0.
libc ABIs: UNIQUE ABSOLUTE
For bug reporting instructions, please see:
<https://bugs.launchpad.net/ubuntu/+source/glibc/+bugs>.
physimo@RaspberryPi:~/Desktop/llama$ 
```

**Reply by giladgd** [reply]
This is indeed the case.
You can see that the `glibc` version that ships with your OS is `2.31`, but the prebuilt binaries need at least version `2.32`. From the logs:
```
1108 error Error: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /home/physimo/Desktop/llama/node_modules/@node-llama-cpp/linux-arm64/bins/linux-arm64/llama-addon.node)
```
(I don't recommend trying to update `glibc` specifically as it could break the entire OS)

The OS is also too old for the newer cpp features that it failed to build.

I recommend you to update to the latest Ubuntu LTS version (24.04) and then reinstall the latest nodejs version directly from nodesource, it should solve this.

**Reply by physimo** [reply]
I see.
I guess a fresh OS install is in order, will have to backup a lot before I could though. This will take a while...

Thanks for the clarification @giladgd , I will tell you the result later whether it works or not, hopefully it's the former.

**Reply by physimo** [reply]
@giladgd It worked!
Thank you very much. I'll close this discussion since the problem has been resolved.
